Aa/js create metadata (#815)
clerk/clerk-docs: docs/references/javascript/user/create-metadata.mdx

---
title: Create User Metadata
description: Explore methods on the User object that help you create data for the user, such as createEmailAddress() and createPhoneNumber().
---

# Create user metadata

These methods on the [`User`](/docs/references/javascript/user/user) object help you create data for the user, such as `createEmailAddress()` and `createPhoneNumber()`.

## `createEmailAddress()`

Adds an email address for the user. A new [`EmailAddress`][email-ref] will be created and associated with the user.

<Callout type="warning">
  **Email address** must be enabled as an identifier in your Clerk settings for this method to work. See the [Identifiers](/docs/authentication/configuration/sign-in-sign-up-options#identifiers) section to learn more.
</Callout>

```typescript
function createEmailAddress: (params: CreateEmailAddressParams) => Promise<EmailAddress>;
```

### `CreateEmailAddressParams`

| Name | Type | Description |
| --- | --- | --- |
| `email` | `string` | The value of the email address |

### `createEmailAddress()` returns

| Type | Description |
| --- | --- |

> | <code>Promise\<[EmailAddress][email-ref]\></code> | This method returns a `Promise` which resolves to the the email address created for the user. |

Reviewer: ```suggestion
| <code>Promise\<[EmailAddress][email-ref]\></code> | This method returns a `Promise` which resolves to the email address created for the user. |
```

### `createEmailAddress()` example

> The following example adds `test@test.com` as an email address for the user. If the user is not signed in, the user will be prompted to sign in.

Reviewer: ```suggestion
The following example adds `test@test.com` as an email address for the user. If the user is signed out, they will be prompted to sign in.
```

  document.getElementById("app").innerHTML = `
    <div id="sign-in"></div>
  `;

  const signInDiv = document.getElementById("sign-in");

  clerk.mountSignIn(signInDiv);
}
```

```html filename="index.html"
  <div id="app"></div>

  <!-- Initialize Clerk with your
  Clerk publishable key and Frontend API URL -->
  <script
    async
    crossorigin="anonymous"
    data-clerk-publishable-key='{{pub_key}}'
    src="https://[your-domain].clerk.accounts.dev/npm/@clerk/clerk-js@latest/dist/clerk.browser.js"
    type="text/javascript"
  ></script>

  <script>
    window.addEventListener("load", async function () {
      await Clerk.load();

      if (Clerk.user) {
        const email = "test@test.com";

        Clerk.user.createEmailAddress({ email })
          .then((response) => console.log(response))
          .catch((error) => console.log("An error occurred:", error.errors));
      } else {
        document.getElementById("app").innerHTML = `
            <div id="sign-in"></div>
          `;

        const signInDiv = document.getElementById("sign-in");

        Clerk.mountSignIn(signInDiv);
      }
    });
  </script>
```
</CodeBlockTabs>

</InjectKeys>

## `createPhoneNumber()`

> Adds a phone number for the user. A new [`PhoneNumber`][phone-ref] will be created and associated with the user.

Reviewer: ```suggestion
This example creates a new [`PhoneNumber`][phone-ref] and associates it with the user.
```

        Clerk.mountSignIn(signInDiv);
      }
    });
  </script>
```
</CodeBlockTabs>

</InjectKeys>

## `createExternalAccount()`

Adds an external account for the user. A new [`ExternalAccount`][exacc-ref] will be created and associated with the user. This method is useful if you want to allow an already signed-in user to connect their account with an external provider, such as Facebook, GitHub, etc., so that they can sign in with that provider in the future.

<Callout type="warning">
  The social provider that you want to connect to must be enabled in your Clerk settings for this method to work. See the [Social connections (OAuth)](/docs/authentication/configuration/sign-in-sign-up-options#social-connections-o-auth) section to learn more.
</Callout>

```typescript
function createExternalAccount: (params: CreateExternalAccountParams) => Promise<ExternalAccount>;
```

### `CreateExternalAccountParams`

| Name | Type | Description |
| --- | --- | --- |
| `strategy` | [`OAuthStrategy`](/docs/references/javascript/types/oauth#o-auth-strategy) | The strategy corresponding to the OAuth provider. For example: `oauth_facebook`, `oauth_github`, etc. |

> | `redirectUrl?` | `string` | The URL to redirect back to one the OAuth flow has completed successfully or unsuccessfully. |

Reviewer: ```suggestion
| `redirectUrl?` | `string` | The URL to redirect back to once the OAuth flow has completed successfully or unsuccessfully. |
```

| `additionalScopes?` | `string[]` | Any additional scopes you would like your user to be prompted to approve. |

### `createExternalAccount()` returns

| Type | Description |
| --- | --- |
| `Promise<ExternalAccount>` | This method returns a `Promise` which resolves to the [`ExternalAccount`][exacc-ref] created for the user. |

### `createExternalAccount()` example

After calling `createExternalAccount`, the initial `state` of the returned `ExternalAccount` will be `unverified`. To initiate the connection with the external provider, redirect the user to the `externalAccount.verification.externalVerificationRedirectURL` contained in the result of `createExternalAccount`.

Upon return, inspect within the `user.externalAccounts` the entry that corresponds to the requested strategy:

> - If the connection was successful, then `externalAccount.verification.status` should be `verified`.

Reviewer: ```suggestion
- If the connection succeeded, then `externalAccount.verification.status` will be `verified`.
```

I prefer "will" because it's more definite. "Should" seems to imply that the connection can succeed and the status can still be "unverified". If that's the case, we definitely gotta document that, but if not we should say "will".

> - If the connection was not successful, then the `externalAccount.verification.status` should not be `verified` and the `externalAccount.verification.error` will contain the error encountered so that you can present corresponding feedback to the user.

Reviewer: ```suggestion
- If the connection was not successful, then the `externalAccount.verification.status` will be `unverified` and the `externalAccount.verification.error` will contain the error encountered, which you can present to the user.
```

Author: I think the status will be `failed` if the verification fails. the status can also be `unverified`.

<img width="680" alt="Screenshot 2024-03-18 at 17 54 58" src="https://github.com/clerk/clerk-docs/assets/98043211/7d5c09ed-8eff-499e-8cef-16e2794c2161">

Reviewer: Got it. Your original wording is the most accurate way to describe it I think. Although I prefer "will not be" over "should not be", as it leaves less room for confusion. I assume it's not expected behavior that an unsuccessful connection will sometimes be `verified`. If that is expected, we'll need to explain why.

> The following example demonstrates how to add a Notion account as an external account for the user. When the user clicks the "Add Notion as a social connection" button, the user will be redirected to Notion to connect their account. After connecting their account, they will be redirected back to the `/` route of your application and the status of the connection will be displayed.

Reviewer: ```suggestion
The following example demonstrates how to add a Notion account as an external account for the user. When the user selects the "Add Notion as a social connection" button, the user will be redirected to Notion to connect their account. After connecting their account, they will be redirected to the `/` route of your application, and the status of the connection will be displayed.
```
